### PR TITLE
Extra Trees Implementation

### PR DESCRIPTION
**Pull request by mtandon18**
**Description**
This PR adds support for Extremely Randomized Trees (or Extra Trees for short) in linfa-trees. Here is a [link](https://quantdare.com/what-is-the-difference-between-extra-trees-and-random-forest/) detailing what Extra Trees are. For convenienve, below is a summary of the main features:

- Many trees w/all input observations used.
- Decision nodes consider only a subset of features and a random split for each feature. The best random split is assigned to the node.

**Primary Changes in Code**
1. Updated the way fitting works in decision trees. First, not all features are considered at every decision node when max_features is not None. Additionally, I have broken out the code that finds the best split across all features and when the hyperparameter random_split is enabled, I skip finding the optimal split for a given feature. These latter changes were just incorporated into the existing for loop since there was very little additional code here.

2. Added fitting and prediction traits for ExtraTrees as well as the hyperparameters ExtraTreesParams. Fitting works by creating as many decision trees trees as specified by num_estimators. Predicting works by passing the observation through all the trees and picking the most frequent prediction.

**Next Steps**
This is untested, except for one doc test (which passes). I will need to add a suite of tests and examples in `extra_trees.rs` in a follow up PR.

### CODE REVIEW

File: algorithms/linfa-trees/examples/extra_trees.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+fn main() {}
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/examples/extra_trees.rs]: this example should actually contain an example or be removed

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+impl<'a, F: Float, L: Label + 'a + std::fmt::Debug, D, T> Fit<ArrayBase<D, Ix2>, T, Error>
+    for ExtraTreesParams<F, L>
+where
+    D: Data<Elem = F>,
+    T: AsTargets<Elem = L> + Labels<Elem = L>,
+{
+    type Object = ExtraTrees<F, L>;
+
+    /// Fit extremely randomized trees using `hyperparamters` on the dataset consisting of
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/algorithm.rs]: what exactly is `hyperparameters` here?

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        let (records, feature_names) = (dataset.records(), dataset.feature_names());
+
+        // Overrides the `max_features` hyperparameter once the dataset is provided
+        // to the square root of the number of features if it is `None`
+        let true_max_features = match self.max_features {
+            None => (num_features as f64).sqrt() as usize,
+            Some(x) => x,
+        };
+
+        let mut all_trees = Vec::new();
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/algorithm.rs]: you can pre-allocate this vec with `Vec::with_capacity`, since we are provided `num_estimators`

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs [hunk 4]
```diff
@@ -510,6 +575,19 @@ impl<F: Float, L: Label + Default, D: Data<Elem = F>> PredictInplace<ArrayBase<D
...
+    feature_names: &'a Vec<String>,
+) -> (RowMask, Vec<SortedIndex<'a, F>>) {
+    let all_idxs = RowMask::all(records.nrows());
+    let sorted_indices: Vec<SortedIndex<F>> = (0..(records.ncols()))
+        .map(|feature_idx| {
+            SortedIndex::of_array_column(records, feature_idx, &feature_names[feature_idx])
+        })
+        .collect();
+    (all_idxs, sorted_indices)
+}
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/decision_trees/algorithm.rs]: I'm not super sure at a glance what exactly this is doing, I think it could benefit from a doc comment. Also, is this actually setting up trees? It looks like it's actually applying some pre-processing to the dataset.

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs [hunk 5]
```diff
@@ -644,10 +716,14 @@ impl<F: Float, L: Label + std::fmt::Debug> DecisionTree<F, L> {
     pub fn export_to_tikz(&self) -> Tikz<F, L> {
         Tikz::new(self)
     }
+
+    pub fn root_node_by_val(self) -> TreeNode<F, L> {
+        self.root_node
+    }
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/decision_trees/algorithm.rs]: I'm not sure I understand the purpose/use case for this function. Was this from a previous iteration of the design? If it's still necessary I would suggest renaming this to `into_root_node` to follow conventions from [here](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv).

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+impl<F: Float, L: Label + std::fmt::Debug> ExtraTrees<F, L> {
+    /// Defaults are provided if the optional parameters are not specified:
+    /// * `decision_tree_params = Default parameters for [decision trees](struct.DecisionTreeParams.html) with random_split overriden`
+    /// * `num_estimators = 100`
+    /// * `max_features = None`
+    /// The `max_features` default of `None` will be overriden to the square root of the number of features when a dataset is provided.
+    // Violates the convention that new should return a value of type `Self`
+    #[allow(clippy::new_ret_no_self)]
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/algorithm.rs]: I don't think this violates that convention, does this actually give a warning?

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        self.validate(num_features)?;
+
+        let (records, feature_names) = (dataset.records(), dataset.feature_names());
+
+        // Overrides the `max_features` hyperparameter once the dataset is provided
+        // to the square root of the number of features if it is `None`
+        let true_max_features = match self.max_features {
+            None => (num_features as f64).sqrt() as usize,
+            Some(x) => x,
+        };
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/algorithm.rs]: https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs [hunk 8]
```diff
@@ -224,92 +347,32 @@ impl<F: Float, L: Label + std::fmt::Debug> TreeNode<F, L> {
...
-                let w = weight_on_right_side / total_weight;
-                let score = w * left_score + (1.0 - w) * right_score;
-
-                // Take the midpoint from this value and the next one as split_value
-                split_value = (split_value + sorted_index.sorted_values[i + 1].1) / F::cast(2.0);
+        // Select subset of features to test
+        let mut test_features: HashMap<usize, &SortedIndex<F>> =
+            sorted_indices.iter().enumerate().collect();
+
+        if matches!(max_features, Some(_)) {
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/decision_trees/algorithm.rs]: since you use the internal value you can do
```rust
if let Some(max_features) = max_features
```

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs [hunk 9]
```diff
@@ -196,13 +197,135 @@ impl<F: Float, L: Label + std::fmt::Debug> TreeNode<F, L> {
...
+                || weight_on_left_side < hyperparameters.min_weight_leaf
+            {
+                continue;
+            }
+
+            // Don't need to compute the quality of other arbitrary splits,
+            // will only consider a split past the random split
+            if hyperparameters.random_split && i < random_split_idx {
+                continue;
+            }
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/decision_trees/algorithm.rs]: I don't understand this comment/the logic here

Comment (mustafa-guler) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: Ok I now understand it, making it explicit that  this is to skip all indices until we hit the random split idx would be helpful. Something like
```rust
// Skip indices until we hit `random_split_idx`, since for random splits we only need to consider the randomly selected index
```

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs [hunk 10]
```diff
@@ -196,13 +197,135 @@ impl<F: Float, L: Label + std::fmt::Debug> TreeNode<F, L> {
...
+        // Consider a single random split rather than an optimal split
+        // if `random_split` is enabled.
+        let mut random_split_idx = 0;
+
+        if hyperparameters.random_split {
+            let valid_split_points: Vec<usize> = (0..mask.mask.len())
+                .filter(|&i| mask.mask[sorted_index.sorted_values[i].0])
+                .collect();
+            random_split_idx = rand::thread_rng().gen_range(0..valid_split_points.len());
+        }
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/decision_trees/algorithm.rs]: I assume the point here is to randomly select a valid point to split at, correct? If so, this doesn't seem to do that to me. The rng function [generates a value from the provided range](https://docs.rs/rand/0.8.4/rand/trait.Rng.html#method.gen_range), but this doesn't correspond to one of the valid mask values.

Also, if this is being computed anyway doesn't it make sense to just loop over the valid mask values instead of having a check inside the following for loop?

File: algorithms/linfa-trees/src/extra_trees/hyperparameters.rs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+#[derive(Clone, Copy, Debug)]
+pub struct ExtraTreesParams<F, L> {
+    pub decision_tree_params: DecisionTreeParams<F, L>,
+    pub num_estimators: usize,
+    pub max_features: Option<usize>,
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/hyperparameters.rs]: since `max_features` is used explicitly during the fitting of the constituent trees I think it makes more sense to move max_features to the `DecisionTreeParams`, so there's less magic in this implementation.

File: algorithms/linfa-trees/src/extra_trees/hyperparameters.rs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    /// Checks the correctness of the hyperparameters
+    ///
+    /// ### Panics
+    ///
+    /// If the minimum impurity increase is not greater than zero
+    pub fn validate(&self, num_features: usize) -> Result<()> {
+        if !self.decision_tree_params.random_split {
+            return Err(Error::Parameters(format!(
+                "Extra trees should always use a random split, but random_split is currently set to {}",
+                self.decision_tree_params.random_split
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/extra_trees/hyperparameters.rs]: format shouldn't be necessary here, the random_split in this error is always going to be false, so you can just write `false` inside the string and do `String::from` or `.to_string()`

File: algorithms/linfa-trees/src/lib.rs [hunk 13]
```diff
@@ -1,7 +1,7 @@
 //!
-//! # Decision tree learning
+//! # Tree based learning
```
Comment (mustafa-guler) [inline algorithms/linfa-trees/src/lib.rs]: I think `Decision tree learning` is still valid here, since extra trees and random forests are really just ensembles of decision trees

**Review by mustafa-guler — CHANGES_REQUESTED** [review]

File: algorithms/linfa-trees/examples/extra_trees.rs (hunk 1 above)
Comment (mtandon18) [reply algorithms/linfa-trees/examples/extra_trees.rs]: Yeah, I mentioned in the **Next Steps** above that I was planning on adding testing + examples in another PR. I didn't finish that this week but still wanted to have some code ready for review. I can tack it onto this PR if you'd prefer that though!

Comment (mustafa-guler) [reply algorithms/linfa-trees/examples/extra_trees.rs]: in that case for this one you can just delete this file

File: algorithms/linfa-trees/src/extra_trees/hyperparameters.rs (hunk 11 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/extra_trees/hyperparameters.rs]: Ooh yes, this was actually something I wanted to discuss! I'm not sure if there's a way to circumvent this problem, but in `fit` I don't have write access for any of the provided parameters (which makes sense) because I'm given a shared reference `&self`. But, the default for `max_features` is dependent on the number of features in the `dataset`. So, I ended up passing it in as an argument but stylistically I'm not a fan of that, especially since it never changes again. 

If this move to `DecisionTreeParams` is to avoid the extra argument, then I think it won't actually serve that purpose. I can technically move max_features into `DecisionTreeParams` with little code change but I was wondering if conceptually it should belong there? scikit-learn seems to use it as a parameter for Extra Trees and not Decision Trees.

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs (hunk 10 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: Yeah, you're right. I thought I had switched it to `choose` at some point but it doesn't look like it saved. 

I thought of just looping throguh `valid_split_points`, but hesitated since it would add an additional unnecessary loop for the decision tree use case. But I guess the optimization matters a lot more in the ensemble scenario so I'll switch this.

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs (hunk 2 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/extra_trees/algorithm.rs]: Bad comment inherited from same location in `decision_trees/algorithm.rs` but it's supposed to be in self

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs (hunk 5 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: Yeah, sorry this is old code that I don't need anymore

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs (hunk 6 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/extra_trees/algorithm.rs]: Yea, seems like a bad comment inherited from decision trees. There are no warnings when I remove it

File: algorithms/linfa-trees/src/extra_trees/algorithm.rs (hunk 2 above)
Comment (mustafa-guler) [reply algorithms/linfa-trees/src/extra_trees/algorithm.rs]: oh I see, I think we can change this for ExtraTrees

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs (hunk 10 above)
Comment (mustafa-guler) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: doesn't a normal decision tree just want the valid_split_points as well? It seems like it's doing a check inside the loop and continuing if it's invalid, unless I'm misunderstanding?

File: algorithms/linfa-trees/src/extra_trees/hyperparameters.rs (hunk 11 above)
Comment (mustafa-guler) [reply algorithms/linfa-trees/src/extra_trees/hyperparameters.rs]: My reasoning for moving it to `DecisionTreeParams` is that it's really explicitly a modification to how individual decision trees perform and by keeping it on in `ExtraTreeParams` we prevent an end user from using a single `DecisitionTree` with `max_features` in the same stream-lined way as the other hyperparameters, even though this functionality is supported.

My thinking is:
1. Add `max_features` to `DecisionTreeParams`
2. Keep it in `ExtraTreeParams` as well, and overwrite user-provided `max_features` in the internal `DecisionTreeParams` similar to how random_split is overwritten

We can discuss more today though

File: algorithms/linfa-trees/src/decision_trees/algorithm.rs (hunk 10 above)
Comment (mtandon18) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: Yeah, I was just originally trying to avoid the pass through all observations to filter and then going through the valid points when we could just go through all of the observations and skip the invalid ones in one pass as it was doing before. The valid points were originally only being computed for the extra trees case but now I'm doing it regardless bc it's definitely not worth worrying about the extra loop for one decision tree.

Comment (mustafa-guler) [reply algorithms/linfa-trees/src/decision_trees/algorithm.rs]: Oh I see now. We can avoid changing the loop then, instead just fixing the random index selection
